internal/p2p/peermanager.go

```diff
@@ -513,12 +513,21 @@ func (m *PeerManager) HasDialedMaxPeers() bool {
 // returned peer.
 func (m *PeerManager) DialNext(ctx context.Context) (NodeAddress, error) {
 	for {
-		address, err := m.TryDialNext()
-		if err != nil || (address != NodeAddress{}) {
-			return address, err
+		if address := m.TryDialNext(); (address != NodeAddress{}) {
+			return address, nil
 		}
+
+		if ctx.Err() != nil {
+			return NodeAddress{}, ctx.Err()
+		}
+
+		if !m.HasDialedMaxPeers() {
+			continue
+		}
```

Comment:
We only get here if there are no addresses to dial, so having this if statement means we will just be constantly spinning in a for loop

Reply:
You mean if connection slots are full? It's that and if there are no other peers to dial. I believe the point of the sleeper was basically there's no point continually calling `TryDialNext` until we either have a new peer through PEX, a peer disconnects or we're able to redial a peer so let's wait until then

Reply:
we're definitely seeing nodes gradually shed peer connections and just never try to dial more peers. I think the "avoid dialing peers unless we need to" is being really careful with a relatively inexpensive operation (dialing) at the expense of being able to maintain connections to enough peers to make progress.

Reply:
I understand you hypothesis that `DialNext` might be somehow blocking and thus we stop attempting to dial peers. I'm just not sure that the `dialWaker` is the culprit. 

> "avoid dialing peers unless we need to"

It's not so much this, but "don't look for a new peer to dial, if we know we don't have any"

Reply:
One thing we can try, is to just remove the `dialWaker` entirely and see if that changes anything

Reply:
I'd definitely be down to do that as an experiment. I think this PR is worth merging, and then following up with removing that. 

Frankly I think the larger problem is that we'll try very hard to hit the peers at the top of the ranked list, and struggle to get peers because of our bias to the sort order.

Reply:
> Frankly I think the larger problem is that we'll try very hard to hit the peers at the top of the ranked list, and struggle to get peers because of our bias to the sort order.

I don't think this is too much of a problem because we apply exponential backoff to peers that we failed to dial, and their score is temporarily reduced

Reply:
I think the other PR would mean that we would dial more and this is mostly just cleanup.

Reply:
I don't think this will dial any more unless we had a bug beforehand where we weren't causing `Wake` in a situation where a new dialable peer pops up

Reply:
There is the one case where a failure to dial *wouldn't* lead to us dialing new peers. 

I like the idea of seeing if we can pull out of needing the dialWaker at all.

Reply:
> There is the one case where a failure to dial wouldn't lead to us dialing new peers.

Where is this? As I said, if we can just fix the cases where we actually need to `Wake` instead of falling back to calling `Wake` at a random interval then I think this is a far better solution

Reply:
I removed this and added what I think is the missing wake call.